Fix std::string::npos and add missing std::string members

phoerious opened the pull request:
`libcpp.string.npos` maps to `std::npos`, which is wrong and does not compile.

This PR fixes the bug as discussed in #4268 and also adds a few missing `std:::string` members.

I haven't run the test cases myself yet, so CI may fail. If that's the case, I'll fix them later.

0dminnimda commented:
You can run related tests with `python -m runtests cpp_stl_string$`, so you can see and fix everything locally. To run ci every time you need a maintainer (until your first PR is merged)

When you run the tests and fix everything, if necessary, you can call scoder (`@scoder`) so that he can run ci and review your changes.

phoerious commented:
Had a chance to run the tests locally now. I fixed one c/p error and besides compiler warnings in other parts of the code that I didn't touch, everything passes on my end.

scoder commented on Cython/Includes/libcpp/string.pxd:
```diff
@@ -58,19 +58,23 @@ cdef extern from "<string>" namespace "std" nogil:
+        string& erase(size_t) except +
+        string& erase() except +
 
         char& at(size_t) except +
         char& operator[](size_t)
-        char& front()  # C++11
-        char& back()   # C++11
+        char& front()
+        char& back()
```
This documentation suggests that the comments are correct:
https://www.cplusplus.com/reference/string/string/

phoerious replied:
They are correct, but to_string is also C++11 and wasn't explicitly marked as such. There may even be more unmarked C++11 features. I'm not sure what your stance on backwards compatibility is, but in 2021, C++11 should be the minimum-supported version in most compilers unless you specify otherwise via compiler flags.

scoder replied:
Ok, let's keep them out then to avoid the confusing mix of versioned and unversioned entries.

scoder commented on Cython/Includes/libcpp/string.pxd:
```diff
@@ -5,7 +5,7 @@ from libc.string cimport const_char
 cdef extern from "<string>" namespace "std" nogil:
 
-    size_t npos = -1
+    const size_t npos "std::string::npos"
```
Why did you chose this spelling over the separate `cdef extern` block given in the [ticket](https://github.com/cython/cython/issues/4268#issue-937625219)?

phoerious replied:
Because it's shorter and seems to generate equivalent code.

scoder replied:
I would prefer if Cython knew from the declaration that the name uses a different C++ namespace, 1) because why not and 2) because I can't tell in what way C++ namespace handling will change in the future in Cython, so better declare it correctly right now.

Explicit cnames are more of a last-resort way to get what your C(++) compiler wants. We should instead try to give good examples in the declarations that we ship. Users may end up actually reading them as The Right Way To Do It.

phoerious replied:
Alright, I changed it. I would even prefer if I could declare it as a static member of `std::string`, because that's what it is. But then Cython would translate it as `std::string.npos`, which obviously doesn't work.

scoder reviewed (commented):
Since this (supposedly) fixes the previous implementation, it would be good to also proof it by adding a test.

phoerious commented:
Wait a minute. I DID add tests. Where are they?

phoerious commented:
There you go. Must have accidentally deleted them during the last force push.

phoerious commented:
Also added `pop_back()`, since we were already at it. The interface should be complete now, I believe.

scoder commented on Cython/Includes/libcpp/string.pxd:
```diff
@@ -175,12 +180,37 @@ cdef extern from "<string>" namespace "std" nogil:
-    string to_string(long val)
-    string to_string (long long val)
-    string to_string (unsigned val)
-    string to_string (unsigned long val)
-    string to_string (unsigned long long val)
-    string to_string (float val)
-    string to_string (double val)
-    string to_string (long double val)
+    string to_string(int)
+    string to_string(long)
+    string to_string(long long)
+    string to_string(unsigned)
+    string to_string(unsigned long)
+    string to_string(unsigned long long)
+    string to_string(float)
+    string to_string(double)
+    string to_string(long double)
```
1) I think it's nice to allow keyword arguments by naming the arguments – although, admittedly, it doesn't seem very attractive here, but there's still a slight chance that this breaks someone's code out there.
2) if we're changing these anyway, then we should also add `except +` since the allocation may fail.

phoerious replied:
I removed the parameter names for consistency with the rest of the file. I did not add an exception clause, because none of the other functions that may throw an allocation exception have one. Only explicit exceptions such as out_of_range or so are caught here.

phoerious replied:
BTW these `to_string` functions throw a compiler error about ambiguous signatures if you pass a `size_t` without an explicit cast to `unsigned long`. I am not quite sure why that is and it if is a special Cython quirk. To fix that I contemplated adding an explicit `size_t` overload, but ultimately decided against it, because it's not an official overload from the C++ documentation (although it would work). What's your stance on that?

scoder replied:
> I did not add an exception clause, because none of the other functions that may throw an allocation exception have one.

They can raise, so they'd better have one for safety.

> To fix that I contemplated adding an explicit `size_t` overload, but ultimately decided against it, because it's not an official overload from the C++ documentation (although it would work). What's your stance on that?

Sounds like a helpful addition. Also `ssize_t`. The declarations are meant to make Cython users happy, not the spec.

phoerious replied:
Done.

scoder commented:
> Must have accidentally deleted them [the tests] during the last force push.

It's better to avoid force pushing the changes. I'll squash-merge your PR at the end, also because it seems to qualify for a backport to the next 0.29.x, since it fixes several issues with the `libcpp.string` declarations. So feel free to use context specific commit messages and incremental changes.

phoerious commented:
I added except clauses to all methods that may throw `bad_alloc` (which were a lot) and also added some more missing overloads for `resize()` and `compare()`.

scoder commented on tests/run/cpp_stl_string.pyx:
```diff
@@ -102,6 +102,15 @@ def test_push_back(char *a):
 
+def test_pop_back(char *a):
+    """
+    >>> test_pop_back(b'abc')
+    b'ab'
```
This doesn't work in Py2. I sometimes write test like this:
```suggestion
    >>> test_pop_back(b'abc') == b'ab'  or  test_pop_back(b'abc')
    True
```
(Technically, the test would then still succeed if the function call returned `True`, but then, let's assume that we're all honest developers who wouldn't let that pass through a code review…)

phoerious replied:
Ah, ok. I wrote it that way, because it produces a useful error message if the test fails, which `assert True` does not.

phoerious replied:
Fixed.

scoder replied:
I know, that's exactly why I use the spelling above. If the comparison fails, I can see the actual output.

phoerious replied:
Hmm. Makes sense.

phoerious commented:
Since I noticed that the file was still very mixed in terms of whether methods have parameter names or not, I went through the trouble of adding parameter names to all signatures. That also should avoid breaking anybody's code.

scoder commented:
Very nice, thanks!
